core/org.wso2.carbon.utils/src/main/java/org/wso2/carbon/utils/httpclient5/CustomHostNameVerifier.java

```diff
@@ -0,0 +1,156 @@
+
+    private static CustomHostNameVerifier customHostNameVerifierInstance = null;
+    private static final DefaultHostnameVerifier DEFAULT_HOSTNAME_VERIFIER = new DefaultHostnameVerifier();
+
+    private static final Logger LOG = LoggerFactory.getLogger(CustomHostNameVerifier.class);
+
+    private static final String[] LOCALHOST_SAN = {"::1", "127.0.0.1", "localhost", "localhost.localdomain"};
+
+    private CustomHostNameVerifier() {
+
+    }
+
+    public static CustomHostNameVerifier getInstance() {
+
+        if (customHostNameVerifierInstance == null) {
+            customHostNameVerifierInstance = new CustomHostNameVerifier();
+        }
+        return customHostNameVerifierInstance;
+    }
+
+    @Override
+    public boolean verify(String host, SSLSession session) {
+
+        try {
+            Certificate[] certs = session.getPeerCertificates();
+            X509Certificate x509 = (X509Certificate) certs[0];
+            this.verify(host, x509);
+            return true;
+        } catch (SSLException e) {
+            if (LOG.isDebugEnabled()) {
+                LOG.debug(e.getMessage(), e);
+            }
+            return false;
+        }
+    }
+    
+    @Override
+    public void verify(String host, X509Certificate cert) throws SSLException {
+        
+        try {
+            // Extract subject alternative names.
+            String[] subjectAlternativeNames = extractSubjectAlternativeNames(cert);
+            
+            // Extract CN from subject.
+            String[] commonNames = extractCommonNames(cert);
+            
+            // Merge subject alternative names with localhost alternatives.
+            String[] subjectAlternativeNamesWithLocalhosts =
+                    (String[]) ArrayUtils.addAll(subjectAlternativeNames, LOCALHOST_SAN);
+
+            if (commonNames.length > 0 && !ArrayUtils.contains(subjectAlternativeNames, commonNames[0])) {
+                subjectAlternativeNamesWithLocalhosts =
+                        (String[]) ArrayUtils.add(subjectAlternativeNamesWithLocalhosts, commonNames[0]);
+            }
+            
+            // Check if the host matches any of our accepted names.
+            if (ArrayUtils.contains(subjectAlternativeNamesWithLocalhosts, host)) {
+                return;
+            }
+            
+            // If not in the extended list, use the default verifier.
+            DEFAULT_HOSTNAME_VERIFIER.verify(host, cert);
+        } catch (CertificateParsingException e) {
+            throw new SSLException("Certificate parsing error", e);
+        }
+    }
+
+    private String[] extractSubjectAlternativeNames(X509Certificate cert) throws CertificateParsingException {
+
+        Collection<List<?>> subjectAltNames = cert.getSubjectAlternativeNames();
+        List<String> result = new ArrayList<>();
+        
+        if (subjectAltNames != null) {
+            for (List<?> san : subjectAltNames) {
```

Comment:
Add a comment explaining what's in the `san` list and why you have to take the 0th element.